subsys/net/lib/mqtt_socket/mqtt_encoder.c

```diff
@@ -0,0 +1,781 @@
+ */
+static u32_t mqtt_encode_fixed_header(u8_t message_type, u32_t length,
+				      u8_t **packet)
+{
+	u32_t packet_length = 0xFFFFFFFF;
+
+	if (length <= MQTT_MAX_PAYLOAD_SIZE) {
+		u32_t offset = 1;
+
+		MQTT_TRC("<< MT:0x%02x L:0x%08x", message_type, length);
+		packet_length_encode(length, NULL, &offset);
+
+		MQTT_TRC("Remaining length size = %02x", offset);
+
+		u8_t *mqtt_header = *packet - offset;
+
+		/* Reset offset. */
+		offset = 0;
+		(void)pack_uint8(message_type, MQTT_MAX_PACKET_LENGTH,
+				 mqtt_header, &offset);
+		packet_length_encode(length, mqtt_header, &offset);
+
+		*packet = mqtt_header;
+
+		packet_length = (length + offset);
+	}
+
+	return packet_length;
+}
+
+/**@brief Encodes a string of a zero length.
+ *
+ * @param[in] buffer_len Total size of the buffer on which string will be
+ *                       encoded. This shall not be zero.
+ * @param[out] buffer Buffer where the string is to be encoded.
+ * @param[inout] offset Offset on the buffer where the string is to be encoded.
+ *                      If the procedure is successful, the offset is
+ *                      incremented to point to the next write/pack location on
+ *                      the buffer.
+ *
+ * @retval 0 if the procedure is successful.
+ * @retval -EINVAL if the offset is greater than or equal to the buffer length.
+ */
+static int zero_len_str_encode(u32_t buffer_len, u8_t *const buffer,
+			       u32_t *const offset)
+{
+	return pack_uint16(0x0000, buffer_len, buffer, offset);
+}
+
+/**@brief Encodes and sends messages that contain only message id in
+ *        the variable header.
+ *
+ * @param[in] client Identifies the client for which the procedure is requested.
+ * @param[in] message_type Message type and reserved bit fields.
+ * @param[in] message_id Message id to be encoded in the variable header.
+ * @param[out] packet A pointer to store a pointer to encoded message.
+ * @param[out] packet_length A pointer to store message length.
+ * @retval 0 or an error code indicating a reason for failure.
+ */
+static int mqtt_message_id_only_enc(const struct mqtt_client *client,
+				    u8_t message_type, u16_t message_id,
+				    const u8_t **packet, u32_t *packet_length)
+{
+	int err_code = -ENOTCONN;
+	u32_t offset = 0;
+	u32_t mqtt_packetlen = 0;
+	u8_t *payload;
+
+	/* Message id zero is not permitted by spec. */
+	if (message_id == 0) {
+		return -EINVAL;
+	}
+
+	payload = &client->tx_buf[MQTT_FIXED_HEADER_EXTENDED_SIZE];
+	memset(payload, 0, MQTT_MAX_VARIABLE_HEADER_N_PAYLOAD);
+
+	err_code = pack_uint16(message_id,
+			       MQTT_MAX_VARIABLE_HEADER_N_PAYLOAD,
+			       payload, &offset);
+
+	if (err_code == 0) {
+		mqtt_packetlen = mqtt_encode_fixed_header(message_type,
+							  offset,
+							  &payload);
+
+		*packet_length = mqtt_packetlen;
+		*packet = payload;
+	} else {
+		*packet_length = 0;
+		*packet = NULL;
+	}
+
+	return err_code;
+}
+
+int connect_request_encode(const struct mqtt_client *client,
```

Comment:
Perhaps this function would look better if it simply returned on error, instead of executing all code inside one or multiple "if not error" statement. If you need to run some generic error handling code before returning, you can jump to it using a goto.

Reply:
Sure, I agree, but if you look into other parts of the code you'll see, that it's just how the original library was written - there's such a logic all across the code. So to be consistent, it would require changes all accros the library, and it's IMHO something more than a library port, as you start to mess with logic.
Of course, as now we have some nice test coverage in Zephyr also ported, it might be resonable to rework the code in a way you describe, as there is a chance to catch any potential bugs introduced during rework. And the code would get more "Zephyrish" look.

Reply:
IMHO the internals can be changed a bit later if needed, most important thing is to agree on the MQTT APIs so that we do not need to change them a lot later.